lib/rules/no-this-in-template-only-components.js

```diff
@@ -0,0 +1,80 @@
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+
+const Rule = require('./base');
+
+function message(original, fixed) {
+  return (
+    `Usage of 'this' in path '${original}' is not allowed in a template-only component. ` +
+    `Use '${fixed}' if it is a named argument or create a component.js for this component.`
+  );
+}
+
+function hasContext(hbsFilePath) {
+  const paths = hbsFilePath.split(path.sep);
+  if (paths[0] === 'app' || paths[0] === 'addon') {
+    if (paths[1] === 'templates') {
+      if (paths[2] === 'components') {
+        // classic structure: /app/templates/components/foo.hbs => /app/components/foo.js
+        const moduleName = path.basename(hbsFilePath, '.hbs');
+        const jsFilePath = path.join(
+          paths[0],
+          'components',
+          ...paths.slice(3, -1),
+          `${moduleName}.js`
+        );
```

Comment:
We need to ensure that alternate file types are valid. Can you:

* Add an option to allow `validComponentExtensions: ['.lol', '.js', '.ts']` (and search through that list)
* Default the list of valid extensions to `.js` and `.ts`